### ISSUE TITLE
[Android] PurchaseAsync returns null when payload == ""

### ISSUE THREAD

**Issue by THplusplus** [jamesmontemagno/InAppBillingPlugin#140]
As payload parameter requires a non-null value I simply passed "", because I don't actually require the payload parameter.
However, this results in PurchaseAsync returning null even after a successful purchase. Calling GetPurchasesAsync afterwards will correctly return the purchase.
The reason seems to be that the google api's response omits the developerPayload if it is empty, resulting in the plugin reading developerPayload as null. The plugin internally filters the returned purchases by sku and payload, and as null != "", the plugin won't be able to find the matching purchase, causing it to return null.
If I pass any other non-null value to payload, it works as expected.

Version Number of Plugin: 1.2.4
Device Tested On: Samsung Galaxy Nexus (Android 4.3)
Simulator Tested On: -
Version of VS: VS Community 2017, 15.6.6
Version of Xamarin: Xamarin.Android.Support.* 26.0.2

### Steps to reproduce the Behavior
* call PurchaseAsync passing "" to the payload parameter
* complete the purchase (tested in the test store using a test account)

### Expected Behavior
Task should complete with Result being a valid InAppBillingPurchase object OR raise an ArgumentException in case "" is supposed to be an invalid value for payload parameter.

### Actual Behavior
Task completes with Result == null.

### Code snippet
InAppBillingPurchase purchase = await CrossInAppBilling.Current.PurchaseAsync(productId, ItemType.InAppPurchase, "");

**Comment by jamesmontemagno** [comment]
You have to give it a payload

**Comment by THplusplus** [comment]
Explicitly stating that it requires a non-null-value in documentation without mentioning that it must be non-emtpy as well is bound to lead people into providing invalid payload value, while not giving any meaningful error is extremely confusing. I strongly recommend giving this topic a few more seconds of attention and considering either a solid input validation, or a bugfix (because not returning a successful purchase IS a bug).

**Comment by jamesmontemagno** [comment]
1.2.4.140-beta

**Comment by EmilAlipiev** [comment]
just a quick question here. I went through so many discussion in SO about this payload parameter but I still dont understand what it is exactly for. it was said that in order user to use a purchase on multiple devices, i should use his email (for example) as payload and provide for each purchaseasync.
But as I tested, it is not the case.
For example, I did purchase as below

```
string AppId = Plugin.DeviceInfo.CrossDeviceInfo.Current.GenerateAppId();
                      
var purchase = await current.PurchaseAsync(productId, Plugin.InAppBilling.Abstractions.ItemType.InAppPurchase, "");
```
So everytime CrossDeviceInfo plugin generates a new guid. even using on different device (as long as I use same gmail account) purchase is returned. If I login with different account, it asks me to purchase again. So actually it doesnt matter for Non-Consumable purchases?

**Comment by jamesmontemagno** [comment]
Don't do what you are doing, it should be a hard coded string.

Your developer payload must match.

https://developer.android.com/training/in-app-billing/purchase-iab-products

The fifth argument contains a 'developer payload' string that you can use to send supplemental information about an order (it can be an empty string). If you specify a string value, Google Play returns this string along with the purchase response. Subsequently, when you make queries about this purchase, Google Play returns this string together with the purchase details.

Caution: Don't use the developerPayload field for security validation purposes. This field isn't always available when completing tasks related to In-app Billing. For more information about security best practices, see the In-app Billing Security and Design guide.

**Comment by EmilAlipiev** [comment]
@jamesmontemagno  thank you the information.
btw are you able to run 1.2.4.140-beta ? I have updated from stable and i got error message as ">System.ArgumentException: An item with the same key has already been added."

I uninstalled and cleaned all obj, bin folders and installed again. it didnt change. it looks like that some references got mixed here.

![image](https://user-images.githubusercontent.com/10422347/39260680-84c3884c-48ba-11e8-8ba0-4f08ddf5d5be.png)
